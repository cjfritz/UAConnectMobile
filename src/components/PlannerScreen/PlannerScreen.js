/* eslint-disable no-restricted-globals */
import React, { Component } from 'react';
import { FlatList, View } from 'react-native';
import { connect } from 'react-redux';
import { map, reduce } from 'lodash';
import {
  Container, Content, Text, Button, Grid, Col, Row,
} from 'native-base';
import CourseCard from '../common/CourseCard/CourseCard';
import { plannerFetch } from '../../actions/PlannerActions';
import { LoadingModal } from '../common/LoadingModal/LoadingModal';
import styles from './PlannerScreen.style';
// component to display planner screen
export class PlannerScreen extends Component {
  // show navigation heade with add course option
  static navigationOptions = ({ navigation }) => (
    {
      title: 'Course Planner',
      headerRight: (() => (
        <Text style={ styles.headerButtonText } onPress={ () => navigation.navigate('PlannerCreateComponent') }>
          Add Course
        </Text>
      ))(),
      headerStyle: {
        backgroundColor: 'darkred',
      },
      headerTitleStyle: {
        color: 'white',
        fontSize: 20,
      },
    }
  );

  // on component mount, fetch the data from firebase
  componentWillMount() {
    const { firebasePlannerFetch } = this.props;
    firebasePlannerFetch();
  }

  // render all the necessary course cards
  renderItem = item => {
    const { navigation } = this.props;
    return (
      <CourseCard navigation={ navigation } courseItem={ item } />
    );
  };

  // on reload button press, fetch coures info again from firebase
  onReload = () => {
    const { firebasePlannerFetch } = this.props;
    firebasePlannerFetch();
  };

  // render the list for the course screen to show courses
  renderList = () => {
    const { planner, loading } = this.props;
    if (!loading && planner.length !== 0) {
      return (
        <FlatList
          data={ planner }
          renderItem={ ({ item }) => this.renderItem(item) }
          keyExtractor={ item => item.uid }
        />
      );
    } else if (!loading) {
      return (
        <View style={ styles.reloadView }>
          <Text style={ styles.reloadText }>No Courses Found</Text>
          <Button style={ styles.reloadButton } onPress={ this.onReload }>
            <Text style={ styles.titleText }>
              Reload
            </Text>
          </Button>
        </View>
      );
    }
    return null;
  };

  // calculate average GPA to show in the header
  getAverageGPA = () => {
    const { planner } = this.props;
    let gradeCounter = 0;
    const gradeTotal = reduce(planner, (sum, item) => {
      const grade = Number(item.grade);
      if (!isNaN(grade) && item.grade !== '') {
        gradeCounter += 1;
        return sum + grade;
      } else {
        return sum;
      }
    }, 0);
    return (gradeTotal / gradeCounter).toFixed(2);
  };

  // calculate total units to show in the header
  getTotalUnits = () => {
    const { planner } = this.props;
    const unitTotal = reduce(planner, (sum, item) => {
      const units = Number(item.units);
      if (!isNaN(units) && item.units !== '') {
        return sum + units;
      } else {
        return sum;
      }
    }, 0);
    return unitTotal;
  };

  // render the course planner screen
  render() {
    // desctructure props and data needed to render
    const { loading } = this.props;
    const averageGPA = this.getAverageGPA();
    const totalUnits = this.getTotalUnits();
    return (
      <Container style={ styles.container }>
        <LoadingModal visible={ loading } loadingLabel='Loading...' />
        <View style={ styles.infoHeader }>
          <Grid>
            <Col style={ styles.col }>
              <Row style={ styles.row }>
                <Text style={ styles.titleText }>Average GPA</Text>
              </Row>
              <Row style={ styles.row }>
                <Text style={ styles.bodyText }>{ isNaN(averageGPA) ? 'N/A' : averageGPA }</Text>
              </Row>
            </Col>
            <Col style={ styles.col }>
              <Row style={ styles.row }>
                <Text style={ styles.titleText }>Total Units</Text>
              </Row>
              <Row style={ styles.row }>
                <Text style={ styles.bodyText }>{ isNaN(totalUnits) ? 'N/A' : totalUnits }</Text>
              </Row>
            </Col>
          </Grid>
        </View>
        <Content>
          { this.renderList() }
        </Content>
      </Container>
    );
  }
}
// map action creators to component props
const mapDispatchToProps = dispatch => (
  {
    firebasePlannerFetch: () => dispatch(plannerFetch()),
  }
);
// map app state to component props
const mapStateToProps = state => {
  const { loading, courses } = state.planner;
  const planner = map(courses, (val, uid) => (
    { ...val, uid }
  ));
  return { planner, loading };
};
// export component to be used
export default connect(mapStateToProps, mapDispatchToProps)(PlannerScreen);
