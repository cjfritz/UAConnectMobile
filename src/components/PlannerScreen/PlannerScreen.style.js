import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  bodyText: {
    color: 'white',
    fontSize: 18,
  },
  col: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  container: {
    justifyContent: 'center',
  },
  headerButtonText: {
    color: 'white',
    right: '20%',
  },
  infoHeader: {
    backgroundColor: 'darkred',
    height: '10%',
    width: '100%',
  },
  reloadButton: {
    alignSelf: 'center',
    backgroundColor: 'darkred',
  },
  reloadText: {
    color: 'black',
    fontSize: 18,
  },
  reloadView: {
    alignItems: 'center',
    alignSelf: 'center',
    justifyContent: 'center',
  },
  row: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  titleText: {
    color: 'white',
    fontSize: 20,
  },
});
