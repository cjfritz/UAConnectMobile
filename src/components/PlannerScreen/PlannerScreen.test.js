import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { PlannerScreen } from './PlannerScreen';

describe('<PlannerScreen />', () => {
  it('can render reload components on lost connection', () => {
    const props = {
      planner: [],
      loading: false,
      firebasePlannerFetch: jest.fn(),
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerScreen { ...props } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render list of course cards', () => {
    const props = {
      planner: [
        {
          course: 'testCourse',
          description: 'testDescription',
          grade: 4.0,
          term: 'testTerm',
          units: 4,
          uid: 'someUid',
        },
      ],
      loading: false,
      firebasePlannerFetch: jest.fn(),
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerScreen { ...props } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });
});
