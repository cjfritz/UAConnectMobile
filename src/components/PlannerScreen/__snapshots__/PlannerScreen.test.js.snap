// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlannerScreen /> can render list of course cards 1`] = `
<Styled(Container)
  style={
    Object {
      "justifyContent": "center",
    }
  }
>
  <LoadingModal
    loadingLabel="Loading..."
    visible={false}
  />
  <Component
    style={
      Object {
        "backgroundColor": "darkred",
        "height": "10%",
        "width": "100%",
      }
    }
  >
    <GridNB>
      <ColumnNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 20,
              }
            }
          >
            Average GPA
          </Styled(Text)>
        </RowNB>
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            4.00
          </Styled(Text)>
        </RowNB>
      </ColumnNB>
      <ColumnNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 20,
              }
            }
          >
            Total Units
          </Styled(Text)>
        </RowNB>
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            4
          </Styled(Text)>
        </RowNB>
      </ColumnNB>
    </GridNB>
  </Component>
  <Styled(Content)>
    <FlatList
      data={
        Array [
          Object {
            "course": "testCourse",
            "description": "testDescription",
            "grade": 4,
            "term": "testTerm",
            "uid": "someUid",
            "units": 4,
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerScreen /> can render reload components on lost connection 1`] = `
<Styled(Container)
  style={
    Object {
      "justifyContent": "center",
    }
  }
>
  <LoadingModal
    loadingLabel="Loading..."
    visible={false}
  />
  <Component
    style={
      Object {
        "backgroundColor": "darkred",
        "height": "10%",
        "width": "100%",
      }
    }
  >
    <GridNB>
      <ColumnNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 20,
              }
            }
          >
            Average GPA
          </Styled(Text)>
        </RowNB>
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            N/A
          </Styled(Text)>
        </RowNB>
      </ColumnNB>
      <ColumnNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 20,
              }
            }
          >
            Total Units
          </Styled(Text)>
        </RowNB>
        <RowNB
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            0
          </Styled(Text)>
        </RowNB>
      </ColumnNB>
    </GridNB>
  </Component>
  <Styled(Content)>
    <Component
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
        }
      }
    >
      <Styled(Text)
        style={
          Object {
            "color": "black",
            "fontSize": 18,
          }
        }
      >
        No Courses Found
      </Styled(Text)>
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "darkred",
          }
        }
      >
        <Styled(Text)
          style={
            Object {
              "color": "white",
              "fontSize": 20,
            }
          }
        >
          Reload
        </Styled(Text)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;
